hecate_lexer/src/lib.rs

```diff
@@ -1 +1,225 @@
+                        self.advance();
+                    } else {
+                        break;
+                    }
+                }
+            }
+        }
+
+        let token_start = self.current_pos;
+        let token_type;
+
+        match self.iter.peek().unwrap() {
+            token if token.is_alphabetic() => {
+                token_type = Token::Identifier;
+                self.collect_identifier();
+            }
+            &'_' => {
+                token_type = Token::Identifier;
+                self.collect_identifier();
+            }
+            token if token.is_numeric() => {
+                token_type = Token::Literal;
+                self.collect_literal();
+            }
+            &'!' => {
+                if self.followed_by_equalsign() {
+                    token_type = Token::Operator(Ne);
+                } else {
+                    token_type = Token::Operator(Not);
+                }
+            }
+            &'+' => {
+                if self.followed_by_equalsign() {
+                    token_type = Token::Operator(AddAssign);
+                } else {
+                    token_type = Token::Operator(Plus);
+                }
+            }
+            &'-' => {
+                if self.followed_by_equalsign() {
+                    token_type = Token::Operator(SubAssign);
+                } else if self.iter.peek().unwrap() == &'>' {
+                    self.advance();
+                    token_type = Token::ReturnType;
+                } else {
+                    token_type = Token::Operator(Minus);
+                }
+            }
+            &'*' => {
+                if self.followed_by_equalsign() {
+                    token_type = Token::Operator(MulAssign);
+                } else {
+                    token_type = Token::Operator(Mul);
+                }
+            }
+            &'/' => {
+                if self.followed_by_equalsign() {
+                    token_type = Token::Operator(DivAssign);
+                } else {
+                    token_type = Token::Operator(Div);
+                }
+            }
+            &'>' => {
+                if self.followed_by_equalsign() {
+                    token_type = Token::Operator(Ge);
+                } else {
+                    token_type = Token::Operator(Gt);
+                }
+            }
+            &'<' => {
+                if self.followed_by_equalsign() {
+                    token_type = Token::Operator(Le);
+                } else {
+                    token_type = Token::Operator(Lt);
+                }
+            }
+            &'=' => {
+                if self.followed_by_equalsign() {
+                    token_type = Token::Operator(Eq);
+                } else {
+                    token_type = Token::Assignment;
+                }
+            }
+            &'&' => {
+                if self.iter.peek().unwrap() == &'&' {
```

Comment:
this is for tokens such as `&&` and `||`, where the second char has to be checked once again, i don't know if there is a cleaner way.

Reply:
as far as i can tell `advance` is not being called between the two peeks. on the other hand, `followed_by_equalssign` calls `self.advance();` before peeking again. (could you please verify this with a test case?) Also careful with naming there, f`followed_by_equalssign` doesnt implies a state change (which `.advance` does) so that might need to be renamed.

Reply:
it was indeed not matching the tokens right. this should be fixed by df9b03c5ee3763f92d0edc88ba5b5a9911b04338